File: src/Pindou.Domain/Entities/Creation/Diagram.cs
using SqlSugar;
using Pindou.Domain.Common;

namespace Pindou.Domain.Entities.Creation;

/// <summary>
/// 拼豆图纸表
/// </summary>
[SugarTable("diagrams")]
public class Diagram : UuidEntity
{
    /// <summary>用户ID</summary>
    [SugarColumn(IsNullable = false, Length = 36)]
    public string UserId { get; set; } = string.Empty;

    /// <summary>图纸名称</summary>
    [SugarColumn(Length = 100, IsNullable = false)]
    public string Name { get; set; } = string.Empty;

    /// <summary>状态:draft/completed/beaded</summary>
    [SugarColumn(Length = 20, IsNullable = false, DefaultValue = "'draft'")]
    public string Status { get; set; } = "draft";

    /// <summary>源图片URL</summary>
    [SugarColumn(Length = 500, IsNullable = false)]
    public string SourceImageUrl { get; set; } = string.Empty;

    /// <summary>预览图URL</summary>
    [SugarColumn(Length = 500, IsNullable = true)]
    public string? PreviewUrl { get; set; }

    /// <summary>无格线预览</summary>
    [SugarColumn(Length = 500, IsNullable = true)]
    public string? PreviewNoGridUrl { get; set; }

    /// <summary>底板规格</summary>
    [SugarColumn(Length = 20, IsNullable = false)]
    public string BoardSize { get; set; } = string.Empty;

    /// <summary>总颗数</summary>
    [SugarColumn(IsNullable = false)]
    public int BeadCount { get; set; }

    /// <summary>难度</summary>
    [SugarColumn(Length = 20, IsNullable = false)]
    public string Difficulty { get; set; } = string.Empty;

    /// <summary>风格</summary>
    [SugarColumn(Length = 20, IsNullable = false)]
    public string Style { get; set; } = string.Empty;

    /// <summary>色号数量</summary>
    [SugarColumn(IsNullable = false, DefaultValue = "0")]
    public int TotalColors { get; set; }

    /// <summary>豆子总数</summary>
    [SugarColumn(IsNullable = false, DefaultValue = "0")]
    public int TotalBeads { get; set; }

    /// <summary>标签数组(JSON)</summary>
    [SugarColumn(ColumnDataType = "text", IsNullable = true)]
    public string? Tags { get; set; }

    /// <summary>版本号</summary>
    [SugarColumn(IsNullable = false, DefaultValue = "1")]
    public int Version { get; set; } = 1;

    /// <summary>来源:create/template</summary>
    [SugarColumn(Length = 20, IsNullable = false, DefaultValue = "'create'")]
    public string SourceType { get; set; } = "create";

    /// <summary>来源模板ID</summary>
    [SugarColumn(Length = 36, IsNullable = true)]
    public string? TemplateId { get; set; }
}

/// <summary>
/// 色号信息表
/// </summary>
[SugarTable("color_infos")]
public class ColorInfo : UuidEntity
{
    /// <summary>图纸ID</summary>
    [SugarColumn(IsNullable = false, Length = 36)]
    public string DiagramId { get; set; } = string.Empty;

    /// <summary>颜色序号</summary>
    [SugarColumn(IsNullable = false)]
    public int ColorIndex { get; set; }

    /// <summary>色号(如M01)</summary>
    [SugarColumn(Length = 20, IsNullable = false)]
    public string ColorCode { get; set; } = string.Empty;

    /// <summary>颜色名称</summary>
    [SugarColumn(Length = 50, IsNullable = false)]
    public string ColorName { get; set; } = string.Empty;

    /// <summary>RGB值</summary>
    [SugarColumn(Length = 20, IsNullable = false)]
    public string Rgb { get; set; } = string.Empty;

    /// <summary>颗数</summary>
    [SugarColumn(IsNullable = false)]
    public int BeadCount { get; set; }

    /// <summary>占比</summary>
    [SugarColumn(Length = 10, IsNullable = false)]
    public decimal Percentage { get; set; }

    /// <summary>位置范围</summary>
    [SugarColumn(Length = 50, IsNullable = true)]
    public string? Position { get; set; }
}

/// <summary>
/// 生成任务表
/// </summary>
[SugarTable("diagram_tasks")]
public class DiagramTask : UuidEntity
{
    /// <summary>用户ID</summary>
    [SugarColumn(IsNullable = false, Length = 36)]
    public string UserId { get; set; } = string.Empty;

    /// <summary>关联图纸ID</summary>
    [SugarColumn(Length = 36, IsNullable = true)]
    public string? DiagramId { get; set; }

    /// <summary>状态:pending/processing/completed/failed</summary>
    [SugarColumn(Length = 20, IsNullable = false, DefaultValue = "'pending'")]
    public string Status { get; set; } = "pending";

    /// <summary>进度0-100</summary>
    [SugarColumn(IsNullable = false, DefaultValue = "0")]
    public int Progress { get; set; }

    /// <summary>当前阶段</summary>
    [SugarColumn(Length = 50, IsNullable = true)]
    public string? CurrentStage { get; set; }

    /// <summary>源图片</summary>
    [SugarColumn(Length = 500, IsNullable = false)]
    public string SourceImageUrl { get; set; } = string.Empty;

    /// <summary>生成参数(JSON)</summary>
    [SugarColumn(ColumnDataType = "text", IsNullable = false)]
    public string Params { get; set; } = "{}";

    /// <summary>错误信息</summary>
    [SugarColumn(Length = 500, IsNullable = true)]
    public string? ErrorMessage { get; set; }

    /// <summary>是否同步执行</summary>
    [SugarColumn(IsNullable = false, DefaultValue = "0")]
    public bool IsSync { get; set; }

    /// <summary>完成时间</summary>
    [SugarColumn(IsNullable = true)]
    public DateTime? CompleteTime { get; set; }
}

File: Code/APP/lib/models/diagram.dart
class Diagram {
  final String id;
  final String userId;
  final String name;
  final String status;
  final String sourceImageUrl;
  final String previewUrl;
  final String previewNoGridUrl;
  final DiagramParams params;
  final ColorSummary colorSummary;
  final int version;
  final String createdAt;
  final String updatedAt;

  Diagram({
    required this.id,
    required this.userId,
    required this.name,
    required this.status,
    required this.sourceImageUrl,
    required this.previewUrl,
    required this.previewNoGridUrl,
    required this.params,
    required this.colorSummary,
    required this.version,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Diagram.fromJson(Map<String, dynamic> json) {
    return Diagram(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      name: json['name'] ?? '',
      status: json['status'] ?? 'draft',
      sourceImageUrl: json['source_image_url'] ?? '',
      previewUrl: json['preview_url'] ?? '',
      previewNoGridUrl: json['preview_no_grid_url'] ?? '',
      params: DiagramParams.fromJson(json['params'] ?? {}),
      colorSummary: ColorSummary.fromJson(json['color_summary'] ?? {}),
      version: json['version'] ?? 1,
      createdAt: json['create_time'] ?? '',
      updatedAt: json['update_time'] ?? '',
    );
  }
}

class DiagramParams {
  final String boardSize;
  final int beadCount;
  final String difficulty;
  final String style;

  DiagramParams({
    required this.boardSize,
    required this.beadCount,
    required this.difficulty,
    required this.style,
  });

  factory DiagramParams.fromJson(Map<String, dynamic> json) {
    return DiagramParams(
      boardSize: json['board_size'] ?? '50x50',
      beadCount: json['bead_count'] ?? 5000,
      difficulty: json['difficulty'] ?? 'simple',
      style: json['style'] ?? 'realistic',
    );
  }
}

class ColorSummary {
  final int totalColors;
  final int totalBeads;

  ColorSummary({
    required this.totalColors,
    required this.totalBeads,
  });

  factory ColorSummary.fromJson(Map<String, dynamic> json) {
    return ColorSummary(
      totalColors: json['total_colors'] ?? 0,
      totalBeads: json['total_beads'] ?? 0,
    );
  }
}

class ColorInfo {
  final int index;
  final String code;
  final String name;
  final String rgb;
  final int beadCount;
  final double percentage;
  final String position;

  ColorInfo({
    required this.index,
    required this.code,
    required this.name,
    required this.rgb,
    required this.beadCount,
    required this.percentage,
    required this.position,
  });

  factory ColorInfo.fromJson(Map<String, dynamic> json) {
    return ColorInfo(
      index: json['index'] ?? 0,
      code: json['code'] ?? '',
      name: json['name'] ?? '',
      rgb: json['rgb'] ?? '#000000',
      beadCount: json['bead_count'] ?? 0,
      percentage: (json['percentage'] ?? 0).toDouble(),
      position: json['position'] ?? '',
    );
  }
}

class GenerateTask {
  final String taskId;
  final String status;
  final int estimatedTime;
  final bool isSync;

  GenerateTask({
    required this.taskId,
    required this.status,
    required this.estimatedTime,
    required this.isSync,
  });

  factory GenerateTask.fromJson(Map<String, dynamic> json) {
    return GenerateTask(
      taskId: json['task_id'] ?? '',
      status: json['status'] ?? 'pending',
      estimatedTime: json['estimated_time'] ?? 0,
      isSync: json['is_sync'] ?? false,
    );
  }
}

class TaskProgress {
  final String taskId;
  final String status;
  final int progress;
  final String currentStage;
  final List<TaskStage> stages;
  final String? diagramId;
  final String? errorMessage;

  TaskProgress({
    required this.taskId,
    required this.status,
    required this.progress,
    required this.currentStage,
    required this.stages,
    this.diagramId,
    this.errorMessage,
  });

  factory TaskProgress.fromJson(Map<String, dynamic> json) {
    return TaskProgress(
      taskId: json['task_id'] ?? '',
      status: json['status'] ?? 'pending',
      progress: json['progress'] ?? 0,
      currentStage: json['current_stage'] ?? '',
      stages: (json['stages'] as List<dynamic>?)
          ?.map((s) => TaskStage.fromJson(s))
          .toList() ?? [],
      diagramId: json['diagram_id'],
      errorMessage: json['error_message'],
    );
  }
}

class TaskStage {
  final String name;
  final int progress;
  final bool completed;

  TaskStage({
    required this.name,
    required this.progress,
    required this.completed,
  });

  factory TaskStage.fromJson(Map<String, dynamic> json) {
    return TaskStage(
      name: json['name'] ?? '',
      progress: json['progress'] ?? 0,
      completed: json['completed'] ?? false,
    );
  }
}